### ISSUE TITLE
[feature] CMakeDeps generator: merge cpp_info of different recipes if common `cmake_file_name`

### ISSUE THREAD

**Issue by SpaceIm** [conan-io/conan#10229]
- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

Sometimes, we want to split components of a project in several recipes while the all project shares a unique CMake config file. conan is not robust to this approach, since `CMakeDeps` generator will override the generated config files to keep at the end information of only one recipe.

Therefore I suggest to track common `cmake_file_name` in the dependency graph, and merge targets sharing this `cmake_file_name` in the same CMake config file generared by `CMakeDeps`, regardless of their recipe origin.

Projects which would benefit of this feature (if they were modularized in conan): Boost, Qt, msgpack

**Comment by memsharded** [comment]
I am afraid this would add too much complexity, for a corner use case that has other potential approaches.
The complexity of ``CMakeDeps`` has very quickly become very high, just trying to address all the particular issues of open source projects for ConanCenter exclusively (not for user packages). Even if projects like Boost or Qt were modularized in Conan (ConanCenter will probably only allow it if the upstream projects start to release the different libraries independently), they will have to be treated as separate packages, with different names, and different ``find_package()``. Splitting something into multiple packages, and aiming to find and use them with a single ``find_package()`` is simply too much.

If @lasote agrees with it, I'd say that this will not be possible to be added.

**Comment by lasote** [comment]
I totally agree with @memsharded. The CMakeDeps is already at the limit to be maintainable and we cannot add this kind of corner case that I'm sure will introduce bugs.
